components/board/fake/board.go

```diff
@@ -163,6 +168,10 @@ type Board struct {
 	GPIOPins   map[string]*GPIOPin
 	logger     logging.Logger
 	CloseCount int
+
+	cancelCtx  context.Context
+	cancelFunc func()
+	wg         sync.WaitGroup
```

Comment:
We use this pattern in so many places that I've made a standardized version of it instead. Import "go.viam.com/rdk/utils" and use a `utils.StoppableWorkers` instead of adding these 3 new fields.

(this also has the benefit of capturing any panics in the goroutine you spin up, and not needing to do timing shenanigans with the mutex inside `Close()`.)

```go
struct Board {
    ...
    workers utils.StoppableWorkers
}

func NewBoard() {
    ...
    b := &Board{
        ...
        workers: utils.NewStoppableWorkers(),
    }
}

func (b *Board) StreamTicks() {
    for _, di := range interrupts {
        di := di // See https://go.dev/blog/loopvar-preview to explain this
        b.workers.AddWorkers(func (cancelCtx context.Context) {
            ...
        })
    }
}

func (b *Board) Close() {
    mutex.Lock()
    defer mutex.Unlock()
    b.workers.Stop()
    ...
}
```

Reply:
what if the worker should be listening on two contexts? it would not be sufficient to listen on just the component context or the stream context in this case

Reply:
The worker is welcome to listen to as many contexts as it wants, and to shut down early if some other one is cancelled first. `StoppableWorkers` just prevents us from copying and pasting the same pattern over and over with slight variations that lead to bugs.

Reply:
So, it's roughly this:
```
func (b *Board) StreamTicks(localContext context.Context) {
    b.workers.AddWorkers(func(workersContext context.Context) {
        select {
        <-localContext.Done():
            return // The stream was canceled. We'll stop immediately, and later on, b.Close() will be nearly instantaneous.
        <-workersContext.Done():
            return // We're closing the component. Stop streaming ticks, even if whatever was receiving them expected more.
        }
    })
}
```

Reply:
I see, will try

Reply:
updated